compose: Add user presence circles in mention and pm typeahead.

m-e-l-u-h-a-n opened the pull request:
It adds presence circles, like we have in buddy_list to the
typeahead suggestions that are given for mentioning users in messages.

Fixes: #17138

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** Testing for the changes is done manually in the developement server,
and also by updating frontend tests to address these changes.


**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
<details>
<summary>
User suggestions with presense circles
</summary>
![image](https://user-images.githubusercontent.com/63504956/106134956-44f32300-618d-11eb-8cc6-d612c6a6f2c7.png)
</details>

Also I slightly changed the `border-color` for `user_circle_empty` on hover state because those empty circles were hidden due to blue color as can be seen in the image above. After changing the `border-color` on hover the presense circles could be seen in even when mouse was over those list items.
<details>
  <summary>Visibility of empty circles after changing the color of border</summary>
  
![image](https://user-images.githubusercontent.com/63504956/106134691-e7f76d00-618c-11eb-9185-74357fade57e.png)

</details>


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

m-e-l-u-h-a-n commented:
First image is not visible I think when I open the spoiler, I tried to correct that many times but it is not getting resolved. Please paste that link in another tab to see that image.

m-e-l-u-h-a-n commented:
@amanagr can you please review it.

amanagr commented:
@m-e-l-u-h-a-n Thanks for working on this! This looks very promising!
My only concern is that the presence circle of the selected user is not clearly visible.
<img width="352" alt="Screenshot 2021-01-30 at 4 56 17 PM" src="https://user-images.githubusercontent.com/25124304/106355160-499a1180-631c-11eb-89ef-a2f74aa6a202.png">

Maybe we can make the circle darker if the user is selected?

m-e-l-u-h-a-n commented:
I included changes for this in first commit and also attached a photograph with slightly darkened circle in the spoiler in topmost comment, it looks like I missed the darkening for the first user suggestion in the list which is already selected. I will updated that.

m-e-l-u-h-a-n commented:
I updated the pr so as per these changes I have slightly increased the border color darkness only if the suggestion is active. Here's the gif displaying the current changes:
![ezgif com-crop](https://user-images.githubusercontent.com/63504956/106357655-5e32d580-632d-11eb-9002-a2600ba97e5e.gif)

can you give it look again.

amanagr commented:
Looks good. Please wait for review from @timabbott. Good work!

timabbott commented:
Visually, this looks pretty good; one question is what user groups and bot users look like in the typeahead with this PR?  I'd like to be sure this doesn't mess up alignment or similar.

m-e-l-u-h-a-n commented:
For bot users too this presence circle is visible as for other users:
![Screenshot_2021-02-05 Zulip Dev - Zulip](https://user-images.githubusercontent.com/63504956/107069935-a3ad4200-6808-11eb-8eb9-745023529796.png)

But for user groups are they are presented without any such presence cicle:
![image](https://user-images.githubusercontent.com/63504956/107070232-0f8faa80-6809-11eb-9793-6c9dac494889.png)

Do we need to show it for user groups too or may be shift thet suggestion to little bit left. BTW current way also looks good i guess.

m-e-l-u-h-a-n commented on static/styles/zulip.css:
```diff
@@ -599,6 +599,19 @@ strong {
+                top: 4px;
+                right: 7px;
+                display: inline-block;
+            }
```
@timabbott here I have defined styles only required for positioning of user_presence_circle in typeahead list in order to ensure they are rendered correctly in in all screen sizes.
Style rules defined in `user_circle.css` mostly deal with general style that are concerned with appearence(color, gradient, border etc.) of different classes of user_circle. So those styles are used from their only and are not copied/overidden here. I added a comment above this block explaining why these properties are overriden here.
So this is ready for review again. Please have a look.

timabbott replied:
I see -- this still feels like a remarkable amount of CSS for just positioning of the circle; can you try simplifying it?

m-e-l-u-h-a-n replied:
Ohh I have written each margin separately which makes it look big and I think I can remove `vertical-align:middle` as I added it just to be double sure it is centered vertically. I will fix it now.

m-e-l-u-h-a-n commented:
@timabbott I have updated the pr to address alignment of the user groups and also simplified css required for positioning of presence  circles please have a look.

* typeahead with users, user_groups and bots:
![image](https://user-images.githubusercontent.com/63504956/109261979-c2f72800-7826-11eb-87f9-8025e10682be.png)
* typeahead with all and stream suggestions:
![image](https://user-images.githubusercontent.com/63504956/109262136-ffc31f00-7826-11eb-89b8-f843ff0f13a6.png)

timabbott commented on static/styles/zulip.css:
```diff
@@ -2018,6 +2035,10 @@ div.focused_table {
     text-align: center;
+
+    &.fa-group {
+        margin-left: 8px;
```
I changed this to 9px and added a `top: 3px` here to make the alignment more consistent with that of user mentions.

timabbott commented:
I added a preparatory commit to fix preexisting broken vertical alignment for the avatars.  Thanks @m-e-l-u-h-a-n!

I'm deploying this to chat.zulip.org now; Can you post to #design announcing the change (with a fresh screenshot) and asking for feedback?  There may be some additional styling tweaks worth doing in that context.

m-e-l-u-h-a-n commented:
Thanks for all reviews and merging it. Just wanted to tell about the preexisting vertical alignment of avatar that, I noticed that but then I thought that it was probably intentional.
